### DISCUSSION TITLE
PDF Generation The prefix "fo" for element "fo:block" is not bound.

### DISCUSSION THREAD

**Discussion by loathwine** [planet42/Laika#244] [Q&A] [answered]
Hi, I've been using 0.14.0 for PDF generation and recently tried to upgrade to 0.17.1 (can't go higher because stuck with CE2). I'm not sure if I'm doing something wrong or if there is an issue with the library. I'm using java 11, not sure if that could cause issues. I've tried with laika versions 0.16.1 and 0.17.0 as well but I get the same error. Here is a small example using ZIO.

```scala
      import zio.interop.catz._
      import laika.io.implicits._
      val blocker = Blocker.liftExecutionContext(
        ExecutionContext.fromExecutor(Executors.newCachedThreadPool()),
      )

      val parser: MarkupParser = MarkupParser.of(Markdown).build
      val renderer: Resource[Task, BinaryTreeRenderer[Task]] =
        Renderer.of(PDF).io(blocker).sequential[Task].build

      parser.parse("hello") match {
        case Left(error) => Task.fail(new IllegalArgumentException(s"Could not parse markdown $error"))
        case Right(doc) =>
          (for {
            file <- Managed.makeEffect[File](File.createTempFile("KYC", ".pdf"))(_.delete())
            _    <- Managed.effect(println("Created file"))
            _ <- renderer.toManagedZIO.tapM(
              _.from(DocumentTreeRoot(DocumentTree(doc.path, Seq(doc)))).toFile(file).render,
            )
            _ <- Managed.effect(println("Rendered pdf"))
          } yield file).use(file => Task.succeed(println(file.getAbsolutePath)))
      }
// Output
Created file
[Fatal Error] :3:48: The prefix "fo" for element "fo:block" is not bound.



org.xml.sax.SAXParseException; lineNumber: 3; columnNumber: 48; The prefix "fo" for element "fo:block" is not bound.
javax.xml.transform.TransformerException: org.xml.sax.SAXParseException; lineNumber: 3; columnNumber: 48; The prefix "fo" for element "fo:block" is not bound.
	at org.apache.xalan.transformer.TransformerIdentityImpl.transform(TransformerIdentityImpl.java:502)
	at laika.render.pdf.PDFRenderer.$anonfun$render$16(PDFRenderer.scala:76)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
```

**Comment by jenshalm** [comment, marked as answer]
For parsing you are using a pure parser without theme support, for rendering you are using the effectful renderers. This way the default FO template does not get applied (older versions had a fallback and therefore still worked in that case). but theme support (introduced in 0.16 and containing all default templates) is not part of the pure base parsers/renderers).

You can fix this either by using a transformer instead of a separate parser and renderer or by using an IO-based parser, too (`MarkupParser.of(Markdown).io(blocker).sequential[Task]`).

**Reply by loathwine** [reply]
I see, I tried with a IO-based parser but I still can't get it to work
```scala
      import zio.interop.catz._
      import laika.io.implicits._
      val blocker = Blocker.liftExecutionContext(
        ExecutionContext.fromExecutor(Executors.newCachedThreadPool()),
      )

      val parser: Resource[Task, TreeParser[Task]] =
        MarkupParser.of(Markdown).io(blocker).sequential[Task].build
      val renderer: Resource[Task, BinaryTreeRenderer[Task]] =
        Renderer.of(PDF).io(blocker).sequential[Task].build

      (for {
        parsedTree <- parser.toManagedZIO.mapM(_.fromInput(InputTree[Task].addString("hello", Root)).parse)
        file       <- Managed.makeEffect[File](File.createTempFile("FOO", ".pdf"))(_.delete())
        _          <- Managed.effect(println("Created file"))
        _ <- renderer.toManagedZIO.tapM(
          _.from(parsedTree.root).toFile(file).render,
        )
        _ <- Managed.effect(println("Rendered pdf"))
      } yield file).use(file => Task.succeed(println(file.getAbsolutePath)))

// Output
Created file

org.apache.fop.fo.ValidationException: "fo:flow" is missing child elements. Required content model: marker* (%block;)+ (See position 45:42)
javax.xml.transform.TransformerException: org.apache.fop.fo.ValidationException: "fo:flow" is missing child elements. Required content model: marker* (%block;)+ (See position 45:42)
	at org.apache.xalan.transformer.TransformerIdentityImpl.transform(TransformerIdentityImpl.java:502)
	at laika.render.pdf.PDFRenderer.$anonfun$render$16(PDFRenderer.scala:76)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
```

Any ideas?

**Reply by jenshalm** [reply]
Well, at least you have a different error now, that's always progress... :-)

I have not actually tried your example, but for a quick check: can you use an actual path for the input string, e.g. `Root / "doc.md"` instead of just `Root`. `Root` is roughly equivalent to `/` in a file system, so it does not really make sense to be applied to a document, it's a path for directories.

**Reply by loathwine** [reply]
Yes, that was the problem! It is working now! :pray: However, now the file name "doc.md" is rendered at the top in the PDF. Is there is a simple way to rename it when rendering or skip it completely? I tried renaming it to "doc.pdf" to improve it a bit but then I got errors again.

**Reply by jenshalm** [reply]
Yeah, those are chapter markers, which are derived from the page title. That is normally the first header of each document, but if the document has no header, it'll be the path as a fallback. I assume your real application does not have just "hello" as the content and having the actual title there will be ok?

**Reply by loathwine** [reply]
Aha, I see, the real application was a very simple document without even a heading so that explains why I got the path title when trying on the real one as well. Thanks for the help! :pray:

**Comment by jenshalm** [comment]
@Edvin-san I converted the issue to a discussion, pinging you here as I'm not sure you get a notification for that.
